### DISCUSSION TITLE
Best Practices or Patterns for implementing Realtime Connections in Supabase?

### DISCUSSION THREAD

**Discussion by aidrecabrera** [supabase/supabase#21995] [Questions]
Despite my app being small-scale and having just a single user (and that's me 😄), I was surprised to find that the Realtime Peak Connections usage has already exceeded the limit in the free plan.

![image](https://github.com/supabase/supabase/assets/61798731/6b17d81b-0227-46cb-82a8-9a804b21847e)

1. What could possibly be the cause of this (I know this is a dumb question)?
2. Are there some best practices or preferably patterns for effectively implementing Realtime in my app?
3. How do I manage and maximize Realtime Connections?

Thank you in advance! 👀

**Comment by GaryAustin1** [comment]
Seems like you are doing something wrong.  Typically there is one connection per browser tab and all channels share that connection.
Are you using supabase-js or another supabase client?

**Reply by aidrecabrera** [reply]
Hello! I am using supabase-js client.

I agree that I am doing something wrong. Maybe because a new connection is opened whenever the page (component) is opened?

Here's an example:

I tried doing this, and for all my tables, for instance, for my `customers` table:

```typescript
useEffect(() => {
  const channel = realtimeCustomers(setCustomers);
  return () => {
      channel.unsubscribe();
  };
}, []);
```

```typescript
function realtimeCustomers(setCustomers: any) {
	const handleRecordInserted = (payload: { new: any }) => {
		setCustomers((prev: any) => [...prev, payload.new]);
	};

	const handleRecordUpdated = (payload: { new: any }) => {
		setCustomers((prev: any) =>
			prev.map((item: any) =>
				item.customer_id === payload.new.customer_id ? payload.new : item
			)
		);
	};

	const handleRecordDeleted = (payload: { old: any }) => {
		setCustomers((prev: any) =>
			prev.filter((item: any) => item.customer_id !== payload.old.customer_id)
		);
	};

	const channel = supabase
		.channel("customer-room")
		.on(
			"postgres_changes",
			{ event: "INSERT", schema: "public", table: "customer" },
			handleRecordInserted
		)
		.on(
			"postgres_changes",
			{ event: "UPDATE", schema: "public", table: "customer" },
			handleRecordUpdated
		)
		.on(
			"postgres_changes",
			{ event: "DELETE", schema: "public", table: "customer" },
			handleRecordDeleted
		)
		.subscribe();

	return channel;
}
```

What do you think? Where should I open the channel?

**Reply by GaryAustin1** [reply]
So if that uses the same supabase client (createClient) it should all be 1 connection no matter how many channels.

**Reply by aidrecabrera** [reply]
What I also did was create a new instance of the supabase client in a separate file:

```typescript
export const supabase = createClient<Database>(
	import.meta.env.VITE_SUPABASE_URL ?? "",
	import.meta.env.VITE_SUPABASE_KEY ?? ""
);
```

then I use this everywhere in my app like this:

```typescript
import { supabase } from "./client";
```

Could this be the culprit or is it okay to do this? And could it be that I just reloaded the application 200 times (which created the client connection), which exceeded my limit?

**Reply by GaryAustin1** [reply]
Sorry, I can't answer if that cause multliple clients in the framework you are using.  If you load up 200 tabs or browsers at once then yes that would make 200 connections.  Even if you don't unsubscribe once a tab is closed/replaced it should timeout and be dropped.   The 200 limit is open at the same time not scattered out throughout the day or month.

**Reply by chanmathew** [reply]
Hi @aidrecabrera did you ever figure this out?

I'm running into the same issue as well and it seems the real time subscriptions are somehow eating up all my direct DB connections, even though supposedly all channels share a single connection as @GaryAustin1 suggests.
